pinot-core/src/main/java/com/linkedin/pinot/core/common/DataFetcher.java

```diff
@@ -0,0 +1,190 @@
+ * fetching process, reducing duplicate codes and garbage collection.
+ *
+ * TODO: Support Multi Value Columns
+ */
+public class DataFetcher {
+  private final IndexSegment _indexSegment;
+
+  private final Map<String, DataSource> _columnToDataSourceMap = new HashMap<>();
+  private final Map<String, Dictionary> _columnToDictionaryMap = new HashMap<>();
+  private final Map<String, BlockValSet> _columnToBlockValSetMap = new HashMap<>();
+
+  private int[] _blockDocIdArray;
+  private int _blockLength;
+  private final Set<String> _columnDictIdLoaded = new HashSet<>();
+  private final Set<String> _columnValueLoaded = new HashSet<>();
+  private final Set<String> _columnHashCodeLoaded = new HashSet<>();
+  private final Map<String, int[]> _columnToDictIdArrayMap = new HashMap<>();
+  private final Map<String, double[]> _columnToValueArrayMap = new HashMap<>();
+  private final Map<String, double[]> _columnToHashCodeArrayMap = new HashMap<>();
+
+  /**
+   * Constructor for DataFetcher.
+   *
+   * @param indexSegment index segment.
+   */
+  public DataFetcher(IndexSegment indexSegment) {
+    _indexSegment = indexSegment;
+  }
+
+  /**
+   * Given a column, fetch its data source.
+   *
+   * @param column column name.
+   * @return data source associated with this column.
+   */
+  public DataSource getDataSourceForColumn(String column) {
+    DataSource dataSource = _columnToDataSourceMap.get(column);
+    if (dataSource == null) {
+      dataSource = _indexSegment.getDataSource(column);
+      _columnToDataSourceMap.put(column, dataSource);
+    }
+    return dataSource;
+  }
+
+  /**
+   * Given a column, fetch its dictionary.
+   *
+   * @param column column name.
+   * @return dictionary associated with this column.
+   */
+  public Dictionary getDictionaryForColumn(String column) {
+    Dictionary dictionary = _columnToDictionaryMap.get(column);
+    if (dictionary == null) {
+      dictionary = getDataSourceForColumn(column).getDictionary();
+      _columnToDictionaryMap.put(column, dictionary);
+    }
+    return dictionary;
+  }
+
+  /**
+   * Given a column, fetch its block value set.
+   *
+   * @param column column name.
+   * @return block value set associated with this column.
+   */
+  public BlockValSet getBlockValSetForColumn(String column) {
+    BlockValSet blockValSet = _columnToBlockValSetMap.get(column);
+    if (blockValSet == null) {
+      blockValSet = getDataSourceForColumn(column).getNextBlock().getBlockValueSet();
+      _columnToBlockValSetMap.put(column, blockValSet);
+    }
+    return blockValSet;
+  }
+
+  /**
+   * Init the data fetcher with a block doc id array and block length. This method should be called before fetching data
+   * for any specific block.
+   *
+   * @param blockDocIdArray block doc id array.
+   * @param blockLength length of the block.
+   */
+  public void initNewBlock(int[] blockDocIdArray, int blockLength) {
```

Comment:
Because we want to cache the results for each block, and just return the cached values if already fetched. If we pass this as part of the call, then we cannot make use of the cache.